### PR TITLE
[FIX] sale_coupon: Avoid discounting more than order amount

### PR DESCRIPTION
**Pull request by agr-odoo**
- Create a promotion of discount 10% with coupon (A)
- Create a promotion of fixed discount (50$) with coupon (B)
- Create a sale order of 50$, apply A and then B

Order will have a negative total.
This is due to the check done on the order amount, which needs to be
checked without the current program, to allow correct stacking of
promotions

opw-2410666

### CODE REVIEW

**Comment by agr-odoo** [comment]
@rdeodoo 
Hello,
May I have your opinion on this?

**Comment by agr-odoo** [comment]
updated code:
updating the reward amount was going to apply the whole amount anyway, the current fix change the way the order amount is collected to work in all cases
Added test

**Comment by rdeodoo** [comment]
Hello @agr-odoo 

I quickly had a functional look and this indeed fixed the flow described but the issue remains the same if you reverse the coupon applying order.
So if you apply B then A (instead of A then B), issue is still there

**Comment by agr-odoo** [comment]
@rdeodoo 
I would like to optimize it further but I cannot find a way without breaking other flow. May I have your opinion on the current pr?

**Comment by agr-odoo** [comment]
@nle-odoo

**Review by nle-odoo — APPROVED** [review]
this seems to work well

robodoo delegate+

before merging it would be good to have an opinion on current code from website, but it's good to have the testcase in any case

File: addons/sale_coupon/tests/test_program_numbers.py
Comment (rdeodoo) [inline addons/sale_coupon/tests/test_program_numbers.py:627]: 100

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Just back from holiday

There is still a bug with VAT, see 
image

Note that the other way around of applying promotions seems to work in this case.

Also, any reason not to target 12.0? 100% same code

**Comment by agr-odoo** [comment]
I see the issue but this start to be a little too far fetched. Do we really want to support this?
In your case, after adding the -290$ coupon in the order would remain 10$. What should we display when the user add the -10% coupon?

> Also, any reason not to target 12.0? 100% same code

just forgot, I'll rebase once we reach a solution

**Comment by rdeodoo** [comment]
> Do we really want to support this?

Support VAT? :eyes: 

> What kind of lines should be added to the order when the user add the -10% coupon?

Same as anything else, up to the max amount (being the discount max amount or the remaining total if less that discount), doesnt make more or less sense to have another coupon making the order go below 0 or to have one coupon vat line to make the order below 0.

You may wanna ask someone else advice in bugfix team leaders to confirm we want to fix that, but that's just the basic flow this PR try to solve, 2 (or more) coupons make the order go below 0. Having 2 products with different VAT is not something unusual.

**Comment by rdeodoo** [comment]
> just forgot, I'll rebase once we reach a solution

Just asking, I always tried to fix coupon at his lowest supported version as it is one of the rare modules to have exact same code, but in 12 it is in enterprise so always a bit annoying to create the PR in enterprise and switch to community, up to you even if it would probably be better to keep it that way

**Comment by agr-odoo** [comment]
> > What kind of lines should be added to the order when the user add the -10% coupon?
> 
> Same as anything else, up to the max amount (being the discount max amount or the remaining total if less that discount), doesn't make more or less sense to have another coupon making the order go below 0 or to have one coupon vat line to make the order below 0.
> 

Maybe I was just overthinking this but in the situation you described above you have
|product|tax|total|
|----------|----|------|
|A          |15|  100|
|B          ||  100|
|C          ||  100|
|disc.     || -290|

If we add a line
|product|tax|total|
|----------|----|------|
|coupon|15|  -10|

then in the invoice we'll have a tax line calculated on 90.
To be "more fair" with respect to the tax amount, probably we should have 3 discount lines
|product|tax|total|
|----------|----|------|
|A          |15|  -3.33|
|B          ||  -3.33|
|C          ||  -3.33|

but then we introduce at least the rounding issue, so I started to wonder if we should really support this case, while it could be easy for users to be more conservative while creating promotions

Maybe I'm missing something in the above reasoning, but anyway I agree with you, so I updated the PR with the suggested change and also updated the test to reflect this case

**Comment by rdeodoo** [comment]
I will review deeper on monday to not approve this on a hurry, but I agree with what you described, this is by the way what we already do for a single coupon when there is a max amount.
So it makes even more sense to do it that way. (talking about deducting the discount amount line per line until no more discount amount or order is 0).
See
image

File: addons/sale_coupon/models/sale_order.py
Comment (rdeodoo) [inline addons/sale_coupon/models/sale_order.py:155]: This should be `price_subtotal`.
Depending on the promotion order between the 2, result might be still negative.

### Using `price_total`
Fixed amount first:
image

Discount first:
image



### Using `price_subtotal`, same flows will become:
Fixed amount first:
image


Discount first:
image

**Review by rdeodoo — CHANGES_REQUESTED** [review]
I did some tests and found some more issues.

1. It can still go negative, that's an issue for the enduser that I think can be quickly fixed. See 
image
Using `price_subtotal` should fix this as commented on the code.
2. While the order is 0 (in one case with price_total) or a few $ (with price_subtotal), both are still terribly wrong at an accounting level as the seller will have to pay to taxes an amount he never received: 
image
Any sell order, even if total amount if positive or 0, will actually cost money to the seller.

The amount is less with `price_subtotal`suggestion but still, that's with only 15% vat..
For tests, ideally you should test with a tax of 50% to see inconsistencies more easily.

All of this is because the fixed discount amount is not using VAT as discount does I think.

What I suggest:
1. use `price_subtotal`, adapt test, don't test only order.amount_total but also amount untaxed and taxed, as it doesn't make sense to have a order with a negative amount untaxed and a positif tax amount. Obviously test will test wrong number for now, just add a TODO
2. Open a task / bugfix for a deeper fix, I think it needs some analysis on what should be the outcome of a sale order with multiple product with different VAT + a fixed amount (greater than remaining amount to pay) + discount amount. And for both promotion order apply.

Let me know if something seems wrong in what I said :+1:

**Comment by agr-odoo** [comment]
> * use `price_subtotal`, adapt test, don't test only order.amount_total but also amount untaxed and taxed, as it doesn't make sense to have a order with a negative amount untaxed and a positif tax amount. Obviously test will test wrong number for now, just add a TODO
> * Open a task / bugfix for a deeper fix, I think it needs some analysis on what should be the outcome of a sale order with multiple product with different VAT + a fixed amount (greater than remaining amount to pay) + discount amount. And for both promotion order apply.

Indeed I see the issue, I've updated the PR with your suggestions. 
This needs to be looked in depth, defining the expected behavior in this edge case before changing too much the code. 

I've made an attempt to fix the numbers but it seems to me that the sale coupon code to function properly would need to be too much tax/amount aware, but maybe I'm wrong here and there is a simple route.

**Review by rdeodoo — APPROVED** [review]
As discussed, this is fixing the critical visible issue of the discount making the order go bellow 0.
Thanks for the python tests, always feeling good on `coupon` modules :D

There is still a critical 'invisible' issue to be fixed but that will be done in another PR so this one can already move forward.
The remaining issue is that even if the order doesn't go bellow 0, it it still wrong tax wise, see 
image

That will need a deeper analysis at an accounting level to implement the right fix.
 
 In the meantime, let's r+ this one :-)

**Comment by rdeodoo** [comment]
https://www.odoo.com/web#id=2439816&action=327&model=project.task&view_type=form&cids=1&menu_id=4720, we will see what we wanna do for the remaining part.

**Comment by agr-odoo** [comment]
I'll manually backport for v12
robodoo r+

**Pull request closed without merging**
